### PR TITLE
test: Add modifications in SELinux disabled mode

### PR DESCRIPTION
**Pull request by bachradsusi**
When targeted SELinux policy is installed it should be possible to setup SELinux while disabled and before it's changed to permissive/enforcing.

Related to https://github.com/linux-system-roles/selinux/issues/188

This test is supposed to fail until https://github.com/linux-system-roles/selinux/pull/194 is merged.

Enhancement:

Reason:

Result:

Issue Tracker Tickets (Jira or BZ if any):
https://github.com/linux-system-roles/selinux/issues/188

### CODE REVIEW

File: tests/tests_modifications_with_selinux_disabled.yml
Comment (richm) [inline tests/tests_modifications_with_selinux_disabled.yml:10]: Can you use

```yaml
    - name: Ensure SELinux tool semanage
      include_tasks: set_selinux_variables.yml
```
to ensure semanage is present?

**Comment by bachradsusi** [comment]
I guess I should also test whether modifications were applied.

File: tests/tests_modifications_with_selinux_disabled.yml
Comment (richm) [inline tests/tests_modifications_with_selinux_disabled.yml:61]: should this task have a `when` condition like `when: selinux_mountpoint.stdout != ""` e.g. don't execute the task if there is nothing to be done?  Then you can get rid of the `changed_when: false`

File: tests/tests_modifications_with_selinux_disabled.yml
Comment (richm) [inline tests/tests_modifications_with_selinux_disabled.yml:51]:
```suggestion
      changed_when: true
```
because you are unconditionally changing the state of the system

File: tests/tests_modifications_with_selinux_disabled.yml
Comment (richm) [inline tests/tests_modifications_with_selinux_disabled.yml:55]: this is ok because the command does not change the system state

File: tests/tests_modifications_with_selinux_disabled.yml
Comment (richm) [inline tests/tests_modifications_with_selinux_disabled.yml:88]:
```suggestion
      changed_when: true
```
unless you add a `when` condition as mentioned above

File: tests/tests_modifications_with_selinux_disabled.yml
Comment (richm) [inline tests/tests_modifications_with_selinux_disabled.yml:91]:
```suggestion
      changed_when: true
```

File: tests/tests_modifications_with_selinux_disabled.yml
Comment (bachradsusi) [reply tests/tests_modifications_with_selinux_disabled.yml:10]: `set_selinux_variables.yml` runs 4 time `semanage` command which can take 2 seconds on my vm and information collected by this task is not necessary for the test. As the test is now it doesn't even need `semanage` so it could be dropped completely.

Comment (richm) [reply tests/tests_modifications_with_selinux_disabled.yml:10]: > `set_selinux_variables.yml` runs 4 time `semanage` command which can take 2 seconds on my vm

ok - I guess we need a separate task which just installs `semanage`, and another task to populate those test variables.  We can do that in a subsequent commit.  I don't like having multiple copy/paste of the code to install `semanage` - it makes things like `ANSIBLE_GATHERING=explicit`, rpm-ostree support, snapshot creation, new test development, etc. harder.

> and information collected by this task is not necessary for the test. As the test is now it doesn't even need `semanage` so it could be dropped completely.

ack

File: tests/tests_modifications_with_selinux_disabled.yml
Comment (bachradsusi) [reply tests/tests_modifications_with_selinux_disabled.yml]: `changed_when: true` is still necessary even with `when: selinux_mountpoint.stdout != ""`

https://github.com/linux-system-roles/selinux/actions/runs/6420308259/job/17432108183?pr=201

```

no-changed-when: Commands should not change things if nothing needs doing.
Error: Commands should not change things if nothing needs doing.
tests/tests_modifications_with_selinux_disabled.yml:29 Task/Handler: Umount selinux mountpoint to emulate SELinux disabled system {{ selinux_mountpoint.stdout }}

Read documentation for instructions on how to ignore specific rule violations.

                  Rule Violation Summary                  
 count tag             profile rule associated tags       
     1 no-changed-when shared  command-shell, idempotency 
```

**Comment by bachradsusi** [comment]
[citest]

File: tests/tests_modifications_with_selinux_disabled.yml
Comment (richm) [inline tests/tests_modifications_with_selinux_disabled.yml:53]: did you mean to have a `rescue` block here to catch errors?

Comment (bachradsusi) [reply tests/tests_modifications_with_selinux_disabled.yml:53]: No, error would mean that the test failed.

The `rescue` used in https://github.com/linux-system-roles/selinux/blob/main/tests/tests_selinux_disabled.yml#L73 is necessary as the role task fails/errors when reboot is required but it hides another errors. And this test should expose them.

Comment (bachradsusi) [reply tests/tests_modifications_with_selinux_disabled.yml:53]: And the role settings in this test does not require reboot.

Comment (richm) [reply tests/tests_modifications_with_selinux_disabled.yml:53]: > OTOH, it should probably do a cleanup in `rescue`. Is there a way to report an error but continue with the cleanup steps?

yes - block/rescue/always

```yaml
block:
  taskA
  taskB # may throw an error
  taskC # might not reach here
rescue:
  handle error - check if error is the expected one and not something else
always:
  cleanup taskC if needed
  cleanup taskB if needed
  cleanup taskA
```

**Comment by bachradsusi** [comment]
[citest]

File: tests/tests_modifications_with_selinux_disabled.yml
Comment (richm) [inline tests/tests_modifications_with_selinux_disabled.yml:37]: The test isn't catching errors, it is "Execute the role and cleanup" - not sure if you are expecting any errors, but the test is failing:
```
TASK [linux-system-roles.selinux : Set SELinux file contexts] ******************
task path: /WORKDIR/git-test-modifications-in-disabled37f0mfqq/tests/roles/linux-system-roles.selinux/tasks/main.yml:74
Friday 06 October 2023  21:18:41 +0000 (0:00:00.013)       0:00:12.398 ******** 
failed: [sut] (item={'target': '/tmp/test_dir(/.*)?', 'setype': 'user_home_dir_t', 'ftype': 'd'}) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": {
        "ftype": "d",
        "setype": "user_home_dir_t",
        "target": "/tmp/test_dir(/.*)?"
    }
}

MSG:

SELinux is disabled on this host.
```

Comment (bachradsusi) [reply tests/tests_modifications_with_selinux_disabled.yml:37]: That's expected because https://github.com/linux-system-roles/selinux/pull/194 is not merged yet. The test demonstrates the current problem and should start passing when https://github.com/linux-system-roles/selinux/pull/194 is merged

Comment (bachradsusi) [reply tests/tests_modifications_with_selinux_disabled.yml:37]: `- name: Execute the role and catch errors` should be just `- name: Execute the role`

**Comment by bachradsusi** [comment]
https://github.com/linux-system-roles/selinux/pull/194 is approved.

I could add this test to https://github.com/linux-system-roles/selinux/pull/194 or wait until it's merged and then [citest] again.

**Comment by richm** [comment]
> #194 is approved.
> 
> I could add this test to #194 or wait until it's merged and then [citest] again.

Yes.  I just merged that one - please rebase this one and we can retest

**Comment by bachradsusi** [comment]
[citest]

**Comment by spetrosi** [comment]
[citest bad]

**Comment by richm** [comment]
I wouldn't worry about the rhel 6 issues, and the rhel 7 issue with the all transitions test and reboot seems to be transient - not easy to reproduce - might be an artifact of the CI system, unless you can consistently reproduce it with qemu locally.

**Comment by richm** [comment]
Is this PR ready to merge?  It conflicts with https://github.com/linux-system-roles/selinux/pull/206, and it would be easier for me to rebase on top of this one, than for you to rebase on top of 206

**Comment by bachradsusi** [comment]
It's ready to be merged. I don't plan any changes in this PR. Thanks!

**Review by richm — APPROVED** [review]

**Pull request merged**
